LIPS/lip-25.md

```diff
@@ -0,0 +1,1983 @@
+* @param _priorityExitShareThresholds array of priority exit share thresholds
+* @param _maxDepositsPerBlock array of max deposits per block
+* @param _minDepositBlockDistances array of min deposit block distances
+*/
+function finalizeUpgrade_v2(
+    uint256[] memory _priorityExitShareThresholds,
+    uint256[] memory _maxDepositsPerBlock,
+    uint256[] memory _minDepositBlockDistances
+) external {
+    _checkContractVersion(1);
+
+    uint256 stakingModulesCount = getStakingModulesCount();
+
+    if (stakingModulesCount != _priorityExitShareThresholds.length) {
+        revert ArraysLengthMismatch(stakingModulesCount, _priorityExitShareThresholds.length);
+    }
+
+    if (stakingModulesCount != _maxDepositsPerBlock.length) {
+        revert ArraysLengthMismatch(stakingModulesCount, _maxDepositsPerBlock.length);
+    }
+
+    if (stakingModulesCount != _minDepositBlockDistances.length) {
+        revert ArraysLengthMismatch(stakingModulesCount, _minDepositBlockDistances.length);
+    }
+
+    for (uint256 i; i < stakingModulesCount; ) {
+        StakingModule storage stakingModule = _getStakingModuleByIndex(i);
+        _updateStakingModule(
+            stakingModule,
+            stakingModule.id,
+            stakingModule.stakeShareLimit,
+            _priorityExitShareThresholds[i],
+            stakingModule.stakingModuleFee,
+            stakingModule.treasuryFee,
+            _maxDepositsPerBlock[i],
+            _minDepositBlockDistances[i]
+        );
+
+        unchecked {
+            ++i;
+        }
+    }
+
+    _updateContractVersion(2);
+}
+```
+
+#### 2.3. Keys vetting improvements
+
+Support of new key vetting logic requires implementation of the new `decreaseStakingModuleVettedKeysCountByNodeOperator` external method. Also, the new `STAKING_MODULE_UNVETTING_ROLE` constant should be defined.
+
+```solidity
+bytes32 public constant STAKING_MODULE_UNVETTING_ROLE = keccak256("STAKING_MODULE_UNVETTING_ROLE");
+
+/// @notice decrese vetted signing keys counts per node operator for the staking module with
+/// the specified id.
+///
+/// @param _stakingModuleId The id of the staking modules to be updated.
+/// @param _nodeOperatorIds Ids of the node operators to be updated.
+/// @param _vettedSigningKeysCounts New counts of vetted signing keys for the specified node operators.
+///
+function decreaseStakingModuleVettedKeysCountByNodeOperator(
+    uint256 _stakingModuleId,
+    bytes calldata _nodeOperatorIds,
+    bytes calldata _vettedSigningKeysCounts
+) external onlyRole(STAKING_MODULE_UNVETTING_ROLE) {
+    _checkValidatorsByNodeOperatorReportData(_nodeOperatorIds, _vettedSigningKeysCounts);
+    _getIStakingModuleById(_stakingModuleId).decreaseVettedSigningKeysCount(_nodeOperatorIds, _vettedSigningKeysCounts);
+}
+```
+
+#### 2.4. New target limit modes
+
+The `StackingRouter` contract should support more than two target limit modes. To implement this the following changes are needed:
```

Comment:
Stacking -> Staking (in several places in the document)

Reply:
Fixed, thanks.